roachtest: check against alpha version for tenant-scope certs

stevendanna opened the pull request:
In semver, v22.2.0 is greater than v22.2.0-alpha. Until the 22.2.0
release is cut, all nightly builds will have pre-release versions that
would fail the previous condition.

Here, I've used an alpha version that I believe is the first alpha
version that required tenant-scoped certs.

Note that we can only do this because we happened to already ship the
version that requires tenant-scoped certs.  Otherwise, we would need a
less exact comparison that would be wrong for some 22.2.0 alpha
versions. An alternative to this type of version comparison might be
if the cockroach binary could report its capabilities via some
standard interface.

Fixes #82929

Release note: None

stevendanna committed:
roachtest: check against alpha version for tenant-scope certs

In semver, v22.2.0 is greater than v22.2.0-alpha. Until the 22.2.0
release is cut, all nightly builds will have pre-release versions that
would fail the previous condition.

Here, I've used an alpha version that I believe is the first alpha
version that required tenant-scoped certs.

Note that we can only do this because we happened to already ship the
version that requires tenant-scoped certs.  Otherwise, we would need a
less exact comparison that would be wrong for some 22.2.0 alpha
versions. An alternative to this type of version comparison might be
if the cockroach binary could report its capabilities via some
standard interface.

Release note: None

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/82927)
<!-- Reviewable:end -->

rimadeodhar reviewed (approved):
Thank you for the fix!

> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82927#-:-N4bzq7079nRgAWgYhRf:b-2q0daq)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @stevendanna)



<!-- Sent from Reviewable.io -->

cucaroach commented:
Thanks for looking into this Steven!  Do roachtests need to be backwards compatible?  My inclination would be to just remove the version check.

stevendanna commented:
@cucaroach In this case, it looked to me that this code was called in at least 1 "upgrade" test where the code needed to work against the proceeding version. (Sometimes it is also nice to make tests work on multiple versions if you can so that bisecting problems isn't too hard.). That said, I'll admit that I didn't read that upgrade test code closely to see if perhaps this check would be safe to remove.

stevendanna commented:
There appears to be at least one more bug here that I'm looking at.  With this patch we are definitely regenerating the certs, but I'm still hitting auth errors.

stevendanna committed:
roachtest: refetch certs after regeneration

This is a hacky fix for build failures related to the new
tenant-scoped certificates. We cache a single copy of the certs to
create connections locally, so each time we regenerate them, we also
need to refetch them. Further, since we want to connect to multiple
tenants, when we regenerate the certs we need to know about all
tenants we might connect to.

Release note: None

stevendanna commented:
@rimadeodhar @cucaroach I pushed another commit that gets me a bit further on most.  I started in on pushing more of this down into roachprod itself, but it is a bit of a tangled mess to be honest.  But, even with this, I'm still seeing errors on the upgrade test:

```
sql_runner.go:176,sql_runner.go:205,multitenant_upgrade.go:315,multitenant_upgrade.go:178,multitenant_upgrade.go:35,test_runner.go:884: error executing 'SELECT * FROM foo LIMIT 1': pq: fai
led to connect to n1 at 10.142.0.63:20011: initial connection heartbeat failed: rpc error: code = Unknown desc = client requested node ID 1 doesn't match server node ID 2
```

stevendanna commented:
bors r+

cucaroach commented:
Awesome thanks @stevendanna!  I think I would run that error by the multitenant slack channel.
